### ISSUE TITLE
bench(model): router-cost A/B — O(E log E) sort vs O(E log k) partial-select at E=384

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12992]
# bench(model): router-cost A/B — O(E log E) sort vs O(E log k) partial-select at E=384

```routing
lane: strix
paths: ["internal/model/v41_router_cost_bench_test.go"]
expected_steps: 5
public_issue: anthony-chaudhary/fak#12640
github_issue: 1308
```

## Parent context

Parent: anthony-chaudhary/fak#12640 (public engine track). Private suite umbrella: #1308.
anthony-chaudhary/fak#12640. Related: `ISSUE-DRAFT-bitonic-topk-router-wiring.md`
(bitonic_topk is implemented but unwired).

## Why this is next

The bottleneck study ranks the router second behind expert streaming: "Router does a full
O(E log E) sort over all 384 experts per layer" (`v41_router.go:115-121`,
`v4_router.go:66-73`), while an O(E log k) partial-selection kernel already exists
(`internal/compute/bitonic_topk.go:61`) and is not wired
(`SUBAGENT-BOTTLENECKS-ORCHESTRATION-VS-MOE-2026-09-12.md:76-81`). 40 layers × every token
pay the full sort. But the cost is only worth fixing if it is non-trivial relative to the
expert-stream cost — this leaf measures it before any wiring decision.

## Problem frame

- Centrality: Core - the router is on the critical path of every token and every layer.
- P1: advanced - a measured baseline is the prerequisite to a wiring issue.
- P2: advanced - a paired benchmark at the real E=384/top-6 shape gives a defensible delta.
- P3: preserved - benchmark-only; does not change or wire the router.
- P4: advanced - the receipt records both arms so a future wiring PR has a baseline.

## Current state

`v41Route` allocates `indices` of width `cfg.Experts` (384) and `sort.SliceStable`s the
whole slice to take top-6, per layer, per token. `bitonic_topk.go` offers a partial
selection. No benchmark compares them at E=384.

## Core through-line

Add a benchmark (`testing.B`) at the pinned V4.1 router shape that times the existing
full-sort top-k against a partial-selection implementation at E=384, top-6, on
representative logits; report ns/op and the ratio; emit a small receipt so the delta is
moment-in-time attested. Benchmark only — do not change `v41Route`.

## Working spine

Benchmark at E=384/top-6 -> time full-sort vs partial-select -> assert top-k set equality -> emit ratio receipt -> no production change.

## Gold-plating boundary

No wiring change to `v41Route`/`v4_router.go`; no new bitonic kernel; no numerical-equality
claim beyond top-k set equality on the sampled logits. The wiring decision is a separate
issue (see the existing draft).

## Verifiable Witness

```bash
go -C ..\fak test ./internal/model -run 'RouterCost|RouterBench' -count=1
go -C ..\fak test ./internal/model -bench 'RouterTopK' -benchtime 2s -count=1
go -C ..\fak vet ./internal/model
```

## Witness

The benchmark reports ns/op for full-sort and partial-select at E=384/top-6 and assert the
two produce identical top-k sets on the sampled logits; the ratio is recorded in a
receipt. No production router behaviour changes.

### File:Line Seam

`internal/model/v41_router_cost_bench_test.go:1` (new); subjects
`internal/model/v41_router.go:115-121`, `internal/model/v4_router.go:66-73`,
`internal/compute/bitonic_topk.go:61`.

### Blast Radius & Affected Lanes

- **Primary Lane:** public `fak` `internal/model`.
- **Unaffected Lanes:** router behaviour; compute kernels.

### Quarantined Fallback Mechanism

If partial-select is not numerically identical on all sampled shapes, the benchmark still
reports both timings and marks the equality check `refused` with the divergence recorded —
it never claims equivalence it did not observe.

## Scoped Acceptance Criteria

- [ ] [SW-VERIFIED] Paired benchmark at E=384/top-6 reports ns/op for both arms.
- [ ] [SW-VERIFIED] Top-k set equality asserted; divergence is a typed refusal, not a pass.
- [ ] [SW-VERIFIED] No change to `v41Route`/`v4_router.go` behaviour.
- [ ] [SW-VERIFIED] Ratio recorded in a receipt with run provenance.

## Done condition

The router-cost delta is measured and witnessed; the wiring decision now has a baseline.

## Acceptance gate

Benchmark runs, focused test green, vet clean.

## Closure binding

The resolving commit cites this leaf and carries the `(fak <leaf>)` trailer.

## Likely files

- `internal/model/v41_router_cost_bench_test.go`

## Work estimate
Estimate: 1 points. Uncertainty: operator-supplied estimate; revise through the parent denominator when evidence changes.

## Overall completion contribution
Contribution: 1/13 points.

## Completion standard
production

## Target operating envelope
- paths: >= 1 command

## Witnessed operating envelope
- paths: 1 command

**Comment by anthony-chaudhary** [comment]
Landed on public `fak` trunk.

- Commit: af0ea716cb9215166e7e8b3c550a49c79301a7ca
- File: `internal/model/v41_router_cost_bench_test.go` (new, benchmark-only)
- Trailer: `(fak model)`; `Refs #12992`

Witness (SW-VERIFIED, public `fak`, trunk af0ea716c):

```text
go vet ./internal/model                                                -> clean
go test ./internal/model -run 'RouterCost' -count=1                    -> ok
go test ./internal/model -bench 'V41RouterTopKCostE384' -benchtime 1s -run '^$' -count=1
  reference_full_sort-32   18700 ns/op   3384 B/op   5 allocs/op
  partial_select-32          590.8 ns/op    152 B/op   4 allocs/op
```

Acceptance criteria:
- [x] [SW-VERIFIED] Paired benchmark at E=384/top-6 reports ns/op for both arms.
- [x] [SW-VERIFIED] Top-k set equality asserted on seeded uniform, all-equal, two-tier, and mod-5 tie-heavy fixtures; the reference arm runs with `FAK_V4_BITONIC_TOPK` unset, and degenerate k mirrors the reference width contract.
- [x] [SW-VERIFIED] No change to `v41Route`/`v4_router.go` behaviour (benchmark-only file; neither router file touched).
- [x] [SW-VERIFIED] Ratio recorded in a `fak/router-cost-ab/1` receipt built from the measured ns/op values; no speedup direction asserted.

No production router behaviour changed. No `[HW-WITNESSED]` criterion touched. V4 Flash 0731 control work untouched. This does not close the parent #12640.
